torch/distributions/utils.py

```diff
@@ -75,3 +76,57 @@ def broadcast_all(*values):
         for idx in scalar_idxs:
             values[idx] = torch.Tensor([values[idx]])
     return values
+
+
+def _get_clamping_buffer(tensor):
+    clamp_eps = 1e-6
+    if isinstance(tensor, Variable):
+        tensor = tensor.data
+    if isinstance(tensor, (torch.DoubleTensor, torch.cuda.DoubleTensor)):
+        clamp_eps = 1e-15
+    return clamp_eps
+
+
+def softmax(tensor):
+    """
+    Wrapper around softmax to make it work with both Tensors and Variables.
+    TODO: Remove once https://github.com/pytorch/pytorch/issues/2633 is resolved.
+    """
+    if not isinstance(tensor, Variable):
+        return F.softmax(Variable(tensor), -1).data
+    return F.softmax(tensor, -1)
+
+
+def log_sum_exp(tensor):
+    """
+    Numerically stable implementation for the `LogSumExp` operation. The
+    summing is done along the last dimension.
```

Comment:
nit: Consider either mentioning that the last dimension is kept or adding a `keepdim=False` argument so that it is clear in usage `log_sum_exp(tensor, keepdim=True)` what the output shape is.